### DISCUSSION TITLE
Problems with Light Handlers

### DISCUSSION THREAD

**Discussion by benjamindonnachie** [cmu-sei/GHOSTS#226] [Q&A] [answered]
Many thanks for all the work that has gone into this framework, I'm hoping to use it to simulate user behaviour on a small network.

In an effort to avoid the MS licensing issues, I'm hoping to use the Light Handlers but even after paring right back to the basics, nothing is created.  There doesn't seem to be a sample timeline for the LightHandlers nor any detailed documentation.  Can anyone assist getting it working please?  Many thanks!

Using ghosts-client-x64-v7.0.0 under Windows 11.

timeline.json:
`{
  "TimeLineHandlers": [
    {
      "HandlerType": "LightExcel",
      "Initial": "",
      "UtcTimeOn": "00:00:00",
      "UtcTimeOff": "24.00:00:00",
      "HandlerArgs": {},
      "Loop": true,
      "TimeLineEvents": [
        {
          "Command": "create",
          "CommandArgs": [
            "%homedrive%%homepath%\\Documents\\Ghost"
          ],
          "DelayAfter": 0,
          "DelayBefore": 0
        }
      ],
      "ScheduleType": "Other",
      "Schedule": null
    }
  ]
}`

(As far as I can tell, "Command" doesn't seem to do anything so also tried without "Command":"create" but issue remains)

Results in the following in app.log:
`2023/07/24 17:17:17.224|Ghosts.Client.Program.Run|Initiating ghosts startup - Local time: 18:17:17.2126561 UTC: 17:17:17.2246711
2023/07/24 17:17:17.475|Ghosts.Domain.Code.ClientConfigurationLoader.get_Config|App config loaded successfully: C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\config\application.json
2023/07/24 17:17:17.475|Ghosts.Client.Program.Run|CheckId instantiated with ID: 
2023/07/24 17:17:17.475|Ghosts.Client.Infrastructure.DebugManager.Write|GHOSTS (ghosts:7.0.0.0 [7.0.161.0]) running in production mode
2023/07/24 17:17:17.475|Ghosts.Client.Infrastructure.DebugManager.Write|Installed path: C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0
2023/07/24 17:17:17.475|Ghosts.Client.Infrastructure.DebugManager.Write|Running as Username: benja - WindowsIdentity: LAPTOP-C1MDSMRP\benja
2023/07/24 17:17:17.521|Ghosts.Client.Infrastructure.DebugManager.Write|C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\config\application.json == True
2023/07/24 17:17:17.521|Ghosts.Client.Infrastructure.DebugManager.Write|C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\config\dictionary.json == True
2023/07/24 17:17:17.521|Ghosts.Client.Infrastructure.DebugManager.Write|C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\config\email-content.csv == True
2023/07/24 17:17:17.521|Ghosts.Client.Infrastructure.DebugManager.Write|C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\config\email-reply.csv == True
2023/07/24 17:17:17.521|Ghosts.Client.Infrastructure.DebugManager.Write|C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\config\emails-domain.json == True
2023/07/24 17:17:17.521|Ghosts.Client.Infrastructure.DebugManager.Write|C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\config\emails-outside.json == True
2023/07/24 17:17:17.521|Ghosts.Client.Infrastructure.DebugManager.Write|C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\config\health.json == True
2023/07/24 17:17:17.521|Ghosts.Client.Infrastructure.DebugManager.Write|C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\config\timeline.json == True
2023/07/24 17:17:17.521|Ghosts.Client.Infrastructure.DebugManager.Write|C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\instance\id.json == False
2023/07/24 17:17:17.521|Ghosts.Client.Infrastructure.DebugManager.Write|C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\instance\files-created.log == False
2023/07/24 17:17:17.521|Ghosts.Client.Infrastructure.DebugManager.Write|C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\instance\trackables.json == False
2023/07/24 17:17:17.521|Ghosts.Client.Infrastructure.DebugManager.Write|C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\instance\survey-results.json == False
2023/07/24 17:17:17.530|Ghosts.Client.Infrastructure.DebugManager.Write|C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\logs\clientupdates.log == False
2023/07/24 17:17:17.657|Ghosts.Client.Infrastructure.DebugManager.Write|------------------
2023/07/24 17:17:17.657|Ghosts.Client.Infrastructure.DebugManager.Write|Client Check-in Values:
2023/07/24 17:17:17.657|Ghosts.Client.Infrastructure.DebugManager.Write|Machine.Id = 
2023/07/24 17:17:17.657|Ghosts.Client.Infrastructure.DebugManager.Write|Machine.Name = laptop-c1mdsmrp
2023/07/24 17:17:17.657|Ghosts.Client.Infrastructure.DebugManager.Write|Machine.CurrentUsername = benja
2023/07/24 17:17:17.657|Ghosts.Client.Infrastructure.DebugManager.Write|Machine.Domain = 
2023/07/24 17:17:17.657|Ghosts.Client.Infrastructure.DebugManager.Write|Machine.ClientIp = 169.254.128.144
2023/07/24 17:17:17.657|Ghosts.Client.Infrastructure.DebugManager.Write|Machine.FQDN = laptop-c1mdsmrp
2023/07/24 17:17:17.657|Ghosts.Client.Infrastructure.DebugManager.Write|Machine.Host = laptop-c1mdsmrp
2023/07/24 17:17:17.657|Ghosts.Client.Infrastructure.DebugManager.Write|Machine.IpAddress = 
2023/07/24 17:17:17.657|Ghosts.Client.Infrastructure.DebugManager.Write|Machine.ResolvedHost = laptop-c1mdsmrp
2023/07/24 17:17:17.657|Ghosts.Client.Infrastructure.DebugManager.Write|------------------
2023/07/24 17:17:17.934|Ghosts.Client.Infrastructure.Email.EmailContentManager.Check|Email content loaded successfully with 7891 records found
2023/07/24 17:17:18.465|Ghosts.Client.Program.Run|Configuration.ResourceControl.ManageProcesses = True
2023/07/24 17:17:18.467|Ghosts.Client.Infrastructure.StartupTasks.CleanupProcesses|Running process cleaner...
2023/07/24 17:17:18.467|Ghosts.Domain.Code.TimelineBuilder.GetLocalTimeline|Loading timeline config C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\config\timeline.json
2023/07/24 17:17:18.515|Ghosts.Client.Infrastructure.StartupTasks.CleanupProcesses|Found ghosts pid: 6452
2023/07/24 17:17:18.515|Ghosts.Client.Infrastructure.StartupTasks.CleanupProcesses|Killing ghosts
2023/07/24 17:17:18.534|Ghosts.Client.Infrastructure.StartupTasks.SetStartup|Set startup registry key successfully
2023/07/24 17:17:18.567|Quartz.Impl.StdSchedulerFactory.InitializeProperties|Default Quartz.NET properties loaded from embedded resource file
2023/07/24 17:17:18.617|Quartz.Impl.StdSchedulerFactory.Instantiate|TaskSchedulingThreadPool configured with max concurrency of 10 and TaskScheduler ThreadPoolTaskScheduler.
2023/07/24 17:17:18.628|Quartz.Core.QuartzScheduler..ctor|Initialized Scheduler Signaller of type: Quartz.Core.SchedulerSignalerImpl
2023/07/24 17:17:18.628|Quartz.Logging.LogProviders.NLogLogProvider+NLogLogger.Log|Quartz Scheduler created
2023/07/24 17:17:18.628|Quartz.Simpl.RAMJobStore.Initialize|RAMJobStore initialized.
2023/07/24 17:17:18.628|Quartz.Impl.StdSchedulerFactory.Instantiate|Quartz Scheduler 3.6.0.0 - 'DefaultQuartzScheduler' with instanceId 'NON_CLUSTERED' initialized
2023/07/24 17:17:18.628|Quartz.Impl.StdSchedulerFactory.Instantiate|Using thread pool 'Quartz.Simpl.DefaultThreadPool', size: 10
2023/07/24 17:17:18.628|Quartz.Impl.StdSchedulerFactory.Instantiate|Using job store 'Quartz.Simpl.RAMJobStore', supports persistence: False, clustered: False
2023/07/24 17:17:18.641|Quartz.Core.QuartzScheduler.Start|Scheduler DefaultQuartzScheduler_$_NON_CLUSTERED started.
2023/07/24 17:17:18.641|Ghosts.Client.Program.Run|CheckID: 
2023/07/24 17:17:18.641|Ghosts.Domain.Code.TimelineBuilder.GetLocalTimeline|Loading timeline config C:\Users\benja\Downloads\ghosts-client-x64-v7.0.0\config\timeline.json
2023/07/24 17:17:18.641|Ghosts.Client.TimelineManager.Orchestrator.Run|watching timeline.json
2023/07/24 17:17:18.641|Ghosts.Client.TimelineManager.Orchestrator.Run|Stopfile watcher is starting
2023/07/24 17:17:18.641|Ghosts.Client.Infrastructure.TempFiles.TempFileWatcher|TempFileWatcher loop beginning
2023/07/24 17:17:18.641|Ghosts.Client.TimelineManager.Orchestrator.WhatsInstalled|Outlook is installed: True
2023/07/24 17:17:18.641|Ghosts.Client.TimelineManager.Orchestrator.SafetyNet|SafetyNet loop beginning
2023/07/24 17:17:18.641|Ghosts.Client.TimelineManager.Orchestrator.WhatsInstalled|Word is installed: True
2023/07/24 17:17:18.641|Ghosts.Client.TimelineManager.Orchestrator.WhatsInstalled|Excel is installed: True
2023/07/24 17:17:18.641|Ghosts.Client.TimelineManager.Orchestrator.WhatsInstalled|PowerPoint is installed: True
2023/07/24 17:17:18.641|Quartz.Logging.LogProviders.NLogLogProvider+NLogLogger.Log|Batch acquisition of 0 triggers
2023/07/24 17:17:18.641|Ghosts.Client.TimelineManager.Orchestrator.ThreadLaunchEx|Attempting new thread for: LightExcel
2023/07/24 17:17:18.641|Ghosts.Client.Infrastructure.TempFiles.TempFileWatcher|TempFileWatcher loop ending
2023/07/24 17:17:18.687|Ghosts.Client.TimelineManager.Orchestrator.SafetyNet|SafetyNet loop ending
2023/07/24 17:17:48.225|Quartz.Logging.LogProviders.NLogLogProvider+NLogLogger.Log|Batch acquisition of 0 triggers
2023/07/24 17:18:16.462|Quartz.Logging.LogProviders.NLogLogProvider+NLogLogger.Log|Batch acquisition of 0 triggers
`

C:\Users\benja\Documents\Ghost remains empty:
```
dir %homedrive%%homepath%\Documents\Ghost
 Volume in drive C is Windows
 Volume Serial Number is 3A50-E107

 Directory of C:\Users\benja\Documents\Ghost

24/07/2023  16:52    <DIR>          .
24/07/2023  16:52    <DIR>          ..
               0 File(s)              0 bytes
               2 Dir(s)  135,837,515,776 bytes free
```

**Comment by sei-dupdyke** [comment, marked as answer]
So, these have been on and off again, and it seems were disabled at some point in the default client. I just published v7.0.166.0 of the client, and they are back again. The sample timelines are [here](https://github.com/cmu-sei/GHOSTS/blob/master/src/Ghosts.Client/Sample%20Timelines/LightHandlers.json). HTH.

**Comment by benjamindonnachie** [comment]
Thank you very much Dustin - seems quite happy now!

2023/07/26 19:17:31.798|Ghosts.Client.Handlers.LightHandlers.GetSavePath|Checking directory at C:\Users\benja\Documents\Ghost\client.xlsx
2023/07/26 19:17:31.798|Ghosts.Client.Handlers.LightHandlers.GetSavePath|Checking directory at C:\Users\benja\Documents\Ghost\PEAR5.docx
2023/07/26 19:17:31.798|Ghosts.Client.Handlers.LightHandlers.GetSavePath|Ghosts.Client.Handlers.LightHandlers+LightExcelHandler saving to path - C:\Users\benja\Documents\Ghost\client.xlsx
2023/07/26 19:17:31.798|Ghosts.Client.Handlers.LightHandlers.GetSavePath|Ghosts.Client.Handlers.LightHandlers+LightExcelHandler saving to path - C:\Users\benja\Documents\Ghost\PEAR5.docx
